### ISSUE TITLE
[Merged on GitLab] Add state response to fixture

### ISSUE THREAD

**Issue by sf-chris** [Sportsflare/bass-tube#126]
## Overview
Add fixture state to the fixture response.
See https://www.notion.so/sportsflare/Autofill-League-settlement-using-fixture-state-01d21c8866734e629d6f536c51ebb644

## Approach

- copy the schemas that are used on ingestion, but relax some of the field requirements

## Considerations
- must be able to handle older versions of state in the db. We have opted for the approach of making fields optional in the fixtures response, rather than backfilling the db to be all consistent with the "current" state, as we anticipate the definition of state increasing over time.

*Originally created on GitLab as an MR by leo_sf (https://gitlab.com/sportsflare/real-time/bass-tube/-/merge_requests/49) on 2022-08-01T02:38:51.493Z*

**Comment by sf-chris** [comment]
If invalid state is written to the database, something's hugely wrong (we should never allow that, and we validate it when we write it). Having said that, I don't think this extra validation hurts us yet so lets leave it in. 

Running this branch locally, adding the state on the API response has increased the response time 4x (10ms -> 40ms on a BO5 fixture) and I'm guessing a bunch of that time would be the marshmallow serializers doing their thing. We could potentially cut that in half by removing this validation, but we can wait until it becomes a problem, as it doesn't currently seem to be much of one.

*Originally commented on GitLab by steffansafey on 2022-08-02T04:03:02.432Z*

**Comment by sf-chris** [comment]
I think if we want the validation of the schema, and to satisfy the field requirements then we need to keep it like this. given that we have build all the title specific state to be handled by the schema, I dont think we wanna remove this.

*Originally commented on GitLab by leo_sf on 2022-08-02T03:12:55.048Z*

**Comment by sf-chris** [comment]
yeah i tested it locally by hitting the endpoint for a bunch of fixtures across time and titles, as well as made sure the list endpoint worked.

*Originally commented on GitLab by leo_sf on 2022-08-02T03:12:54.948Z*

**Comment by sf-chris** [comment]
removed

*Originally commented on GitLab by leo_sf on 2022-08-02T03:12:54.864Z*

**Comment by sf-chris** [comment]
> So that the property is predictably there, just `null` if the fixture doesn't have a state
set to `required=True, allow_none=True`

*Originally commented on GitLab by leo_sf on 2022-08-02T03:12:54.699Z*

**Comment by sf-chris** [comment]
done

*Originally commented on GitLab by leo_sf on 2022-08-02T02:57:40.717Z*

**Comment by sf-chris** [comment]
Left a few comments but largely LGTM man - nice turnaround time on it

I guess we could test it locally by hitting the endpoint for each fixture in the db? (and ensuring `200`s are returned)

*Originally commented on GitLab by steffansafey on 2022-08-01T21:42:52.226Z*

**Comment by sf-chris** [comment]
Potential optimisation: Do you think it would be possible to skip this step? (turning the json string into a dataclass, just for it to be immediately turned back into a json response again)

*Originally commented on GitLab by steffansafey on 2022-08-01T21:42:52.145Z*

**Comment by sf-chris** [comment]
This and !48 should be rebased onto master (the branches aren't based on what's in prod) 
![image](/uploads/476b06d8569ef8b4006da0d51a8be2fc/image.png)

*Originally commented on GitLab by steffansafey on 2022-08-01T21:42:52.064Z*

**Comment by sf-chris** [comment]
Unused

*Originally commented on GitLab by steffansafey on 2022-08-01T21:42:51.988Z*

**Comment by sf-chris** [comment]
I reckon we make this `required=True, allow_none=False`?

So that the property is predictably there, just `null` if the fixture doesn't have a state

*Originally commented on GitLab by steffansafey on 2022-08-01T21:42:51.850Z*

**Comment by sf-chris** [comment]
I have opted to make attributes in the dataclass optional as necessary to suit the most relaxed schema (which for now is the fixture response schema)

This is a shared dataclass, but the schema for ingestion, which is the other place this is used, is strict. 

I have offloaded strict validation to the schema, which allows me to reuse this dataclass in both the ingestion schema and the fixture response schema. A more strict approach would be to duplicate the dataclasses as well and have their optional fields exactly match their corresponding schema, but this seemed unecessary IMO.

Happy for you to let me know otherwise @steffansafey

*Originally commented on GitLab by leo_sf on 2022-08-01T02:41:38.648Z*
